src/persist-types/src/txn.rs

```diff
@@ -74,3 +92,23 @@ pub trait TxnsCodec: Debug {
     /// treated the same as `Some(true)`.
     fn should_fetch_part(data_id: &ShardId, stats: &PartStats) -> Option<bool>;
 }
+
+/// [crate::columnar::Schema] of a data shard used with txn-wal.
+pub trait TxnsDataSchema: Debug + PartialEq {
+    /// Encode this schema for permanent storage.
+    ///
+    /// This must perfectly round-trip Self through [TxnsDataSchema::decode]. If
+    /// the encode function for this schema ever changes, decode must be able to
+    /// handle bytes output by all previous versions of encode.
+    fn encode(&self) -> Bytes;
+    /// Decode a schema previous encoded with this schema's
+    /// [TxnsDataSchema::encode].
+    ///
+    /// This must perfectly round-trip Self through [TxnsDataSchema::decode]. If
```

Comment:
s/decode/encode/ Done.

Correct, it does with the exception of empty. We could also use an Option instead to make that more explicit, but I think I didn't do that because having the option in TxnsEntry::Register is really clunky. Should be pretty easy to move the is_empty check up from the impls of `TxnsDataSchema` into txn-wal if you think that's more obvious?

Reply:
I think it's fine as is